src/main/java/org/veupathdb/service/dsdl/Service.java

```diff
@@ -59,8 +60,18 @@ public GetDownloadByProjectAndStudyIdAndReleaseAndFileResponse getDownloadByProj
     String datasetHash = checkPermsAndFetchDatasetHash(studyId, StudyAccess::allowResultsAll);
     Path filePath = new FileStore(Resources.getDatasetsParentDir(project))
         .getFilePath(datasetHash, release, fileName)
-        .orElseThrow(NotFoundException::new);
+        .orElseThrow(() -> {
+          LOG.info("Unable to find a file with hash {}, release {}, and name {}.", datasetHash, release, fileName);
+          return new NotFoundException();
+        });
     String dispositionHeaderValue = "attachment; filename=\"" + fileName + "\"";
+    Optional<String> userId = UserProvider.lookupUser(_request)
+        .map(user -> Long.toString(user.getUserID()));
+    ServiceMetrics.downloadReporter()
```

Comment:
Hi Dan, I know you didn't request me but I looked anyway.  I find this pattern is useful when building an object that contains one or some of an array of optional properties.  It saves e.g. the creation of a large set of constructors with different param combinations.  But in this case, all these values are required to submit the metric AND they are all known at the same time (even at submission time!), so a single method that takes them all is probably preferable.  At minimum you should probably check in the report() method that they are all set.

Reply:
I can add the validation. I really don't like having a method signature with 3 strings. It feels really easy to mix them up and this felt like a fairly low-cost way to make it obvious which parameter is being passed as which value.

Reply:
I can remove the builder if you really think it's overkill. I just can't write unit tests easily for this method, which is the other way I feel good about ensuring arg order doesn't get screwed up